README.md
## Sensu::Extensions::CheckDependencies

This filter extension provides the Sensu Core built-in filter `check_dependencies`.

This filter matches events when an event already exists, enabling the user to
reduce notification noise and only be notified for the “root cause” of a given
failure. Check dependencies can be defined in the check definition, using
dependencies, an array of checks (e.g. `check_app`), Sensu client/check pairs
(e.g. `db-01/check_mysql`), or a subscription/check pair
(e.g. `subscription:mysql/check_mysql`).

[![Build Status](https://travis-ci.org/sensu-extensions/sensu-extensions-check-dependencies.svg?branch=master)](https://travis-ci.org/sensu-extensions/sensu-extensions-check-dependencies)

### Installation

This extension requires Sensu version >= 0.26 and is provided as a built-in 
filter in Sensu >= 0.29.

To manually install this extension on a Sensu server machine:

```
$ sudo sensu-install -e check-dependencies:1.1.0
```

Edit `/etc/sensu/conf.d/extensions.json` to load it:

```
{
  "extensions": {
    "check-dependencies": {
      "version": "1.1.0"
    }
  }
}
```

### Example Check/Handler Configuration

Specify a dependency on the `mysql` check:

``` javascript
{
  "checks": {
    "web_application_api": {
      "command": "check-http.rb -u https://localhost:8080/api/v1/health",
      "subscribers": [
        "web_application"
      ],
      "interval": 20,
      "dependencies": [
        "mysql"
      ]
    }
  }
}
```

... or specify a dependency on another client's `mysql` check:

``` javascript
{
  "checks": {
    "web_application_api": {
      "command": "check-http.rb -u https://localhost:8080/api/v1/health",
      "subscribers": [
        "web_application"
      ],
      "interval": 20,
      "dependencies": [
        "db-01/mysql"
      ]
    }
  }
}
```

... or specify a dependency on any `mysql` check in the `mysql_nodes`
subscription:

``` javascript
{
  "checks": {
    "web_application_api": {
      "command": "check-http.rb -u https://localhost:8080/api/v1/health",
      "subscribers": [
        "web_application"
      ],
      "interval": 20,
      "dependencies": [
        "subscription:mysql_nodes/mysql"
      ]
    }
  }
}
```

Apply the `check_dependencies` filter to one or more handlers:

``` javascript
{
  "handlers": {
    "custom_mailer": {
      "type": "pipe",
      "command": "custom_mailer.rb",
      "filters": ["check_dependencies"]
    }
  }
}
```


CHANGELOG.md
# Change Log for Sensu::Extension::CheckDependencies

This project adheres to [Semantic Versioning](http://semver.org/).

This CHANGELOG follows the format listed at [Keep A Changelog](http://keepachangelog.com/).

## [Unreleased]

### Fixed

### Added

### Changed

## [1.1.0] - 2018-06-21

### Added

- Add support for setting a subscription/check pair as a check dependency.
[PR #12](https://github.com/sensu/sensu-extensions-check-dependencies/pull/12) (@testeddoughnut)

## [1.0.1] - 2017-03-15

### Fixed

- Fixed a file naming error which made 1.0.0 release unusable.

## [1.0.0] - 2017-02-13

### Added

- Initial release.

[Unreleased]: https://github.com/sensu-extensions/sensu-extensions-check-dependencies/compare/v1.1.0...HEAD
[1.1.0]: https://github.com/sensu-extensions/sensu-extensions-check-dependencies/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/sensu-extensions/sensu-extensions-check-dependencies/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/sensu-extensions/sensu-extensions-check-dependencies/compare/v1.0.0...v1.0.0


spec/helpers.rb
require "rspec"
require "webmock/rspec"
require "eventmachine"
require "sensu/logger"
require "sensu/settings"
require "securerandom"

RSpec.configure do |config|
  config.before(:each) do
    stub_request(:get, %r(127\.0\.0\.1:4567/events/.*/.*)).
      to_return(:status => 404)
  end
end

module Helpers
  def timer(delay, &callback)
    periodic_timer = EM::PeriodicTimer.new(delay) do
      callback.call
      periodic_timer.cancel
    end
  end

  def async_wrapper(&callback)
    EM.run do
      timer(10) do
        raise "test timed out"
      end
      callback.call
    end
  end

  def async_done
    EM.stop_event_loop
  end

  def logger
    Sensu::Logger.get(:log_level => :fatal)
  end

  def settings
    Sensu::Settings.get
  end

  def epoch
    Time.now.to_i
  end

  def client_template
    {
      :name => "i-424242",
      :address => "127.0.0.1",
      :subscriptions => [
        "test"
      ],
      :timestamp => epoch
    }
  end

  def check_template
    {
      :name => "test",
      :type => "standard",
      :issued => epoch,
      :command => "echo WARNING && exit 1",
      :subscribers => [
        "test"
      ],
      :interval => 1,
      :output => "WARNING",
      :status => 1,
      :executed => epoch,
      :history => [0, 1]
    }
  end

  def event_template
    one_second_ago = epoch - 1
    {
      :id => ::SecureRandom.uuid,
      :client => client_template,
      :check => check_template,
      :last_ok => one_second_ago,
      :last_state_change => one_second_ago,
      :occurrences => 1,
      :occurrences_watermark => 1,
      :action => :create
    }
  end
end


spec/check_dependencies_spec.rb
require File.join(File.dirname(__FILE__), "helpers")
require "sensu/extensions/check-dependencies"

describe "Sensu::Extension::CheckDependencies" do
  include Helpers

  before do
    @extension = Sensu::Extension::CheckDependencies.new
    @extension.logger = Sensu::Logger.get
    @extension.settings = {}
  end

  it "can provide the extension API" do
    expect(@extension).to respond_to(:name, :description, :definition, :safe_run, :has_key?, :[])
  end

  it "can determine if an event exists for a check dependency" do
    async_wrapper do
      event = event_template
      @extension.safe_run(event) do |output, status|
        expect(status).to eq(1)
        event[:check][:dependencies] = "invalid"
        @extension.safe_run(event) do |output, status|
          expect(status).to eq(1)
          event[:check][:dependencies] = ["foo"]
          @extension.safe_run(event) do |output, status|
            expect(status).to eq(1)
            stub_request(:get, "127.0.0.1:4567/events/i-424242/foo").
              to_return(:status => 200)
            @extension.safe_run(event) do |output, status|
              expect(status).to eq(0)
              async_done
            end
          end
        end
      end
    end
  end

  it "can determine if an event exists for a check dependency client/check pair" do
    async_wrapper do
      event = event_template
      event[:check][:dependencies] = ["foo"]
      @extension.safe_run(event) do |output, status|
        expect(status).to eq(1)
        event[:check][:dependencies] = ["foo", "bar/qux"]
        @extension.safe_run(event) do |output, status|
          expect(status).to eq(1)
          stub_request(:get, "127.0.0.1:4567/events/i-424242/qux").
            to_return(:status => 200)
          @extension.safe_run(event) do |output, status|
            expect(status).to eq(1)
            stub_request(:get, "127.0.0.1:4567/events/bar/qux").
              to_return(:status => 200)
            @extension.safe_run(event) do |output, status|
              expect(status).to eq(0)
              async_done
            end
          end
        end
      end
    end
  end

  it "can determine if an event exists for a check dependency subscription/check pair" do
    async_wrapper do
      event = event_template
      event2 = event_template
      event[:check][:dependencies] = ["subscription:foo/bar"]
      stub_request(:get, "127.0.0.1:4567/events").
        to_return(:status => 200, :body => JSON.dump([event2]))
      @extension.safe_run(event) do |output, status|
        expect(status).to eq(1)
        event2[:client][:subscriptions] = ["foo"]
        event2[:check][:name] = "bar"
        stub_request(:get, "127.0.0.1:4567/events").
          to_return(:status => 200, :body => JSON.dump([event2]))
        @extension.safe_run(event) do |output, status|
          expect(status).to eq(0)
          async_done
        end
      end
    end
  end
end


lib/sensu/extensions/check-dependencies.rb
require "sensu/extension"
require "timeout"
require "net/http"

module Sensu
  module Extension
    class CheckDependencies < Filter
      def name
        "check_dependencies"
      end

      def description
        "filter events when an event exists for a check dependency"
      end

      # Make an HTTP GET request to the Sensu API, using the URI
      # path provided. Uses Sensu settings to determine how to
      # connect to the API.
      #
      # @param path [String]
      # @return [Object] http response object.
      def sensu_api_get_request(path)
        api = @settings[:api] || {}
        request = Net::HTTP::Get.new(path)
        if api[:user]
          request.basic_auth(api[:user], api[:password])
        end
        Net::HTTP.new(api[:host] || '127.0.0.1', api[:port] || 4567).start do |http|
          http.request(request)
        end
      end

      # Check to see if an event exists for a client/check pair. This
      # method is looking for a HTTP response code of `200`.
      #
      # @param client_name [String]
      # @param check_name [String]
      # @return [Boolean]
      def client_event_exists?(client_name, check_name)
        path = "/events/#{client_name}/#{check_name}"
        response = sensu_api_get_request(path)
        response.code.to_i == 200
      end

      # Check to see if an event exists for a subscription/check pair. This
      # method is looking for a HTTP response code of `200`.
      #
      # @param subscription_name [String]
      # @param check_name [String]
      # @return [Boolean]
      def subscription_event_exists?(subscription_name, check_name)
        path = "/events"
        response = sensu_api_get_request(path)
        events = JSON.load(response.body)
        !events.select { |event| event[:client][:subscriptions].include?(subscription_name) && event[:check][:name] == check_name}.empty?
      end

      # Determine if an event exists for any of the check
      # dependencies declared in the event data, specified in array,
      # check `dependencies`. A check dependency can be a check
      # executed by the same Sensu client (eg. `check_app`), a
      # client/check pair (eg.`i-424242/check_mysql`), or a
      # subscription/check pair (eg. `subscription:mysql/check_mysql`).
      #
      # @param event [Hash]
      # @return [Boolean]
      def dependency_events_exist?(event)
        if event[:check][:dependencies].is_a?(Array)
          event[:check][:dependencies].any? do |dependency|
            begin
              check_name, entity = dependency.split("/").reverse
              if entity =~ /^subscription:.*$/
                subscription_name = entity.split(":")[1]
                subscription_event_exists?(subscription_name, check_name)
              else
                client_name = entity
                client_name ||= event[:client][:name]
                client_event_exists?(client_name, check_name)
              end
            rescue => error
              @logger.error("failed to query api for a check dependency event", :error => error)
              false
            end
          end
        else
          false
        end
      end

      def run(event, &callback)
        filter = Proc.new do
          begin
            Timeout::timeout(10) do
              if dependency_events_exist?(event)
                ["event exists for check dependency", 0]
              else
                ["no current events for check dependencies", 1]
              end
            end
          rescue => error
            @logger.error("check dependencies filter error", :error => error.to_s)
            ["check dependencies filter error: #{error}", 1]
          end
        end
        EM.defer(filter, callback)
      end
    end
  end
end
